cubed/array_api/dtypes.py

```diff
@@ -86,3 +87,35 @@
     "complex floating-point": _complex_floating_dtypes,
     "floating-point": _floating_dtypes,
 }
+
+
+# A Cubed-specific utility.
+def _validate_and_define_dtype(x, dtype=None, *, allowed_dtypes=("numeric",), fname=None, device=None):
+    """Ensure the input dtype is allowed. If it's None, provide a good default dtype."""
+    dtypes = __array_namespace_info__().default_dtypes(device=device)
+
+    # Validate.
+    is_invalid = all(x.dtype not in _dtype_categories[a] for a in allowed_dtypes)
+    if is_invalid:
+        errmsg = f"Only {' or '.join(allowed_dtypes)} dtypes are allowed"
+        if fname:
+            errmsg += f" in {fname}"
+        raise TypeError(errmsg)
+
+    # Choose a good default dtype, when None
+    if dtype is None:
+        if x.dtype in _boolean_dtypes:
+            dtype = dtypes["integral"]
+        elif x.dtype in _signed_integer_dtypes:
+            dtype = dtypes["integral"]
+        elif x.dtype in _unsigned_integer_dtypes:
+            # Type arithemetic to produce an unsinged integer dtype at the same default precision.
+            dtype = nxp.dtype(dtypes["integral"].str.replace("i", "u"))
```

Comment:
According to the spec, dtypes are not required to have a `str` attribute, or a str constructor - so this may not work for all implementations. However, I don't know another way of doing it!

Reply:
Really good point! I found a way to do it using this function: 
https://numpy.org/doc/2.1/reference/generated/numpy.iinfo.html